packages/core/base/src/adapter.ts

```diff
@@ -28,6 +37,7 @@ export interface WalletAdapterProps<Name extends string = string> {
     publicKey: PublicKey | null;
     connecting: boolean;
     connected: boolean;
+    supportedTransactionVersions: Set<TransactionVersion> | null;
```

Comment:
We could type this such that if `supportedTransactionVersions` is null, `sendTransaction`, `signTransaction`, and `signAllTransactions` would only accept the legacy `Transaction` type. This enforces that the caller needs to check the version before they call it, which is similar to what they have to do now (in checking if the methods for `signTransaction` and `signAllTransactions` are defined).

Reply:
Cool, that was my intent. Just to be clear, you weren't suggesting any changes here right?

Reply:
Ah, I'm proposing that this would change to something like

```ts
interface WalletAdapterProps<Name extends string = string, SupportedTransactionVersions extends Set<TransactionVersion> | null = null> {
    supportedTransactionVersions: SupportedTransactionVersions;
    sendTransaction: SupportedTransactionVersions extends null
        ? (transaction: Transaction, ...) => Promise<TransactionSignature>
        : (transaction: Transaction | VersionedTransaction, ...) => Promise<TransactionSignature>
}
```

Now the caller has type safety and can't call it with `VersionedTransaction` until they check for `supportedTransactionVersions`.

Reply:
Oh, I misunderstood your comment, thanks for the clarification!

Reply:
I wonder if it's possible to do this more simply... I don't know if this has the same effect?
```ts
interface WalletAdapterProps<Name extends string = string> {
    supportedTransactionVersions: Set<number> | null;
    sendTransaction: this['supportedTransactionVersions'] extends null
        ? (
              transaction: Transaction,
              connection: Connection,
              options?: SendTransactionOptions
          ) => Promise<TransactionSignature>
        : (
              transaction: Transaction | number,
              connection: Connection,
              options?: SendTransactionOptions
          ) => Promise<TransactionSignature>;
}
```

Reply:
I wonder if it's possible to do this more simply... I don't know if this has the same effect?
```ts
interface WalletAdapterProps<Name extends string = string> {
    supportedTransactionVersions: Set<TransactionVersion> | null;
    sendTransaction: this['supportedTransactionVersions'] extends null
        ? (
              transaction: Transaction,
              connection: Connection,
              options?: SendTransactionOptions
          ) => Promise<TransactionSignature>
        : (
              transaction: Transaction | VersionedTransaction,
              connection: Connection,
              options?: SendTransactionOptions
          ) => Promise<TransactionSignature>;
}
```

Reply:
If this can work (with generics or not), this feels pretty clean:

```ts
export type TransactionOrVersionedTransaction<SupportedTransactionVersions extends Set<TransactionVersion> | null> =
    SupportedTransactionVersions extends null ? Transaction : (Transaction | VersionedTransaction);

export interface WalletAdapterProps<Name extends string = string> {
    name: WalletName<Name>;
    // ...
    supportedTransactionVersions: Set<TransactionVersion> | null;
    sendTransaction(
        transaction: TransactionOrVersionedTransaction<this['supportedTransactionVersions']>,
        connection: Connection,
        options?: SendTransactionOptions
    ): Promise<TransactionSignature>;
}

export interface SignerWalletAdapterProps<Name extends string = string> extends WalletAdapterProps<Name> {
    signTransaction<T extends TransactionOrVersionedTransaction<this['supportedTransactionVersions']>>(
        transaction: T
    ): Promise<T>;
    signAllTransactions<T extends TransactionOrVersionedTransaction<this['supportedTransactionVersions']>[]>(
        transaction: T
    ): Promise<T>;
}
```

Speaking of which, it's probably time for the generic `Name` to go! It doesn't actually do anything useful (because you don't have type safety for which wallets are available inside of your React context, which is how this is mostly used).